modules/dxp/apps/commerce/commerce-organization-web/src/main/resources/META-INF/resources/js/utils/fetch.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * The contents of this file are subject to the terms of the Liferay Enterprise
+ * Subscription License ("License"). You may not use this file except in
+ * compliance with the License. You can obtain a copy of the License by
+ * contacting Liferay, Inc. See the License for the specific language governing
+ * permissions and limitations under the License, including but not limited to
+ * distribution rights of the Software.
+ */
+
+import {fetch, openToast} from 'frontend-js-web';
+
+export function fetchFromHeadless(
+	url,
+	params = {},
+	successMessage,
+	showErrorMessage
+) {
+	return fetch(url, params)
+		.then(handleFetchResponse)
+		.then((data) => handleResponseOk(data, successMessage))
+		.catch((error) => handleResponseNotOk(error, showErrorMessage));
+}
+
+export function handleFetchResponse(response) {
+	return response
+		.json()
+		.then((data) =>
+			response.ok ? Promise.resolve(data) : Promise.reject(data)
+		)
+		.catch((error) =>
+			response.ok
+				? Promise.resolve(null)
+				: Promise.reject({...error, title: error.status})
+		);
```

Comment:
The `fetch` within `fetchFromHeadless` goes to `catch` every time the response is empty even if `response.ok` is true. This function at line 39 kinda `flatten` the results so `handleResponseOk` and `handleResponseNotOk` receive consistent values.

Anyway it's worth to test it live against the api (I wrote it months ago - this chart got blocked for 3 months - and I might have missed/forget something)

Reply:
It looks mostly ok, but there are some weird-looking cases I am not sure are correct. For example, on L34 we have an error that still resolves as a success with no data. It's also unusual for me to see `than`/`catch` blocks within a `then` block. Is this really necessary?

My main point is this can be likely simplified, with https://github.com/liferay-frontend/liferay-portal/pull/1311#discussion_r682510343, unless there is some case I'm not seeing, that code in this PR covers.

Reply:
Hi @marcoscv-work, I tried to replace the code above. Here's a quick summary

|   	|  `response.ok` | `!response.ok`  |
|---	|---	|---	|
|   json | ✅|  ❓ |
|   empty response | ❌ |   ✅|

The ❓stands for those times where I need to show an error message that comes via API in a JSON format.
In the example, the error message is fixed based on FE assumptions. If I need to parse the response to get the info, I need to handle the message error in a `then` after calling the `.json` method.

The  ❌ happens when the response is fine but the body cannot be parsed (there are some POST which return an empty body so it's ok) so they get caught as errors: `Unexpected end of JSON input`

Reply:
This sounds like a bad API... can we fix it?

Reply:
So... a `POST` API should return the entity... but nevertheless I think you should check the `response.ok` first and then use the json or not depending if there is something in the ok/error.

Reply:
@nhpatt the POST requests had already been fixed in the meantime by @luca-pellizon (<3).

They're on a separate [PR](https://github.com/liferay-usm/liferay-portal/pull/189) and get used every time we connect a user to an organization/account.

Still, a DELETE must (or should/can - not sure about be best practices) always return an empty body and that function can handle it as well : )

Reply:
Yep, the `DELETE` should not return anything and a status code of 204 (No Content) but that's why I mentioned checking .ok first, because you avoid the call to .json() (you don't avoid it if you want to get the error message, yeah).